cleanrl/ppo_dna_atari_envpool.py

```diff
@@ -0,0 +1,474 @@
+    torch.nn.init.orthogonal_(layer.weight, std)
+    torch.nn.init.constant_(layer.bias, bias_const)
+    return layer
+
+
+class Agent(nn.Module):
+    def __init__(self, envs):
+        super().__init__()
+        self.register_buffer("obs_mean", torch.zeros(envs.single_observation_space.shape))
+        self.register_buffer("obs_std", torch.full(envs.single_observation_space.shape, 255.0))
+        self.network = nn.Sequential(
+            layer_init(nn.Conv2d(4, 32, 8, stride=4)),
+            nn.ReLU(),
+            layer_init(nn.Conv2d(32, 64, 4, stride=2)),
+            nn.ReLU(),
+            layer_init(nn.Conv2d(64, 64, 3, stride=1)),
+            nn.ReLU(),
+            nn.Flatten(),
+            layer_init(nn.Linear(64 * 7 * 7, 512)),
+            nn.ReLU(),
+        )
+        self.actor = layer_init(nn.Linear(512, envs.single_action_space.n), std=0.01)
+        self.critic = layer_init(nn.Linear(512, 1), std=1)
+
+    def normalize_obs(self, x):
+        x = (x - self.obs_mean.unsqueeze(0)) / (self.obs_std.unsqueeze(0) + 1e-8)
+        return torch.clip(x, -3, 3)
+
+    def get_value(self, x):
+        return self.critic(self.network(self.normalize_obs(x))).squeeze(-1)
+
+    def get_action(self, x, action=None):
+        probs, _ = self.get_action_distribution_and_value(x)
+        if action is None:
+            action = probs.sample()
+        return action, probs.log_prob(action), probs.entropy()
+
+    def get_action_distribution_and_value(self, x):
+        hidden = self.network(self.normalize_obs(x))
+        logits = self.actor(hidden)
+        probs = Categorical(logits=logits)
+        value = self.critic(hidden).squeeze(-1)
+        return probs, value
+
+
+def main():
+    args = parse_args()
+    run_name = f"{args.env_id}__{args.exp_name}__{args.seed}__{int(time.time())}"
+    if args.track:
+        import wandb
+
+        wandb.init(
+            project=args.wandb_project_name,
+            entity=args.wandb_entity,
+            sync_tensorboard=True,
+            config=vars(args),
+            name=run_name,
+            monitor_gym=True,
+            save_code=True,
+        )
+    writer = SummaryWriter(f"runs/{run_name}")
+    writer.add_text(
+        "hyperparameters",
+        "|param|value|\n|-|-|\n%s" % ("\n".join([f"|{key}|{value}|" for key, value in vars(args).items()])),
+    )
+
+    # TRY NOT TO MODIFY: seeding
+    random.seed(args.seed)
+    np.random.seed(args.seed)
+    torch.manual_seed(args.seed)
+    torch.backends.cudnn.deterministic = args.torch_deterministic
+
+    device = torch.device("cuda" if torch.cuda.is_available() and args.cuda else "cpu")
+
+    # env setup
+    envs = envpool.make(
+        args.env_id,
+        env_type="gym",
+        num_envs=args.num_envs,
+        # batch_size=args.epoch_size,  # async
+        num_threads=args.envpool_num_threads,
+        episodic_life=True,
+        reward_clip=True,
+        seed=args.seed,
```

Comment:
Just a note. We designed DNA to run on the original rewards, and it works best with reward_clip off.  That being said, this would mean enabling reward normalization (i.e. dividing by the variance of returns).

Reply:
yeah I was actually thinking about the same thing, and maybe that could explain (part of) the differences in the atari-5 repro results.

Reply:
I removed the clipping and added reward normalization